### ISSUE TITLE
MSI package issues

### ISSUE THREAD

**Issue by GitMensch** [VSCodium/vscodium#830]
At least in `VSCodium-x64-updates-disabled-1.60.0.msi` the language is set to/detected as "Chinese (Traditional, Taiwan)". So "ignore language" option needs to be set in software deployment systems that detect the language and compare it to Windows language.

_Originally posted by @bluikko in https://github.com/VSCodium/vscodium/issues/106#issuecomment-915814070_

**Comment by daiyam** [comment]
`zh-tw` is the last language added so it seems to become the default.

**Comment by GitMensch** [comment]
I bet there's a configuration option to explicit set it (or best: activate "try to find the one matching the OS"), I suggest we see if someone knows of that and otherwise check the wix-docs in some days.

**Comment by daiyam** [comment]
@daiyam In my test, the language were English. Maybe it's something about the deployment systems...

**Comment by bluikko** [comment]
Yes, it is about deployment systems. This example is from Active Directory Software Installation - the most dumbest one. If run manually I guess it could just install the one matching the OS language but AD SI seems to always just pick one language and unless the language is set to be ignored it won't install anything (I don't have a Chinese OS available, I guess that would install OK).

**Comment by daiyam** [comment]
@bluikko Do you know what we will have to improve that?

**Comment by bluikko** [comment]
> @bluikko Do you know what we will have to improve that?

I do not know. I can test if someone can build an MSI following your idea about the last language added being the default.

**Comment by GitMensch** [comment]
I'm quite sure that this will work - but that it won't be the correct solution. There has to be a configuration option for that...

**Comment by bluikko** [comment]
> I'm quite sure that this will work - but that it won't be the correct solution. There has to be a configuration option for that...

I suspect it is a peculiarity of Microsoft's AD SI.

If the MSI is installed manually the language will be (?) chosen to match the OS (as @daiyam reported in https://github.com/VSCodium/vscodium/issues/830#issuecomment-915839346).

AD SI will just take the first language it sees (or some configured language) and installation is not done.

But, if "ignore language" option is selected in AD SI, then the MSI is installed and installed language is correct English (for English OS).

**Comment by InterLinked1** [comment]
I can confirm that I just tried adding this as well and it defaults to Chinese - Traditional (Tawian). Otherwise, though, MSI seems to work great!

**Comment by alexhass** [comment]
The language is auto-detected in this installer. That is normal behaviour Honestly I was not aware that an undetected language ends with chinese... that sounds like a wrong choice to me, too. I would default to english.

It is also normal that you need to enable ignore language in AD deployment if the MSI is mulitlingual.

This can be solved by removing all languages from the language field - except 1033 - english. It is also an option that the project builds an installer for English only. Personally I like this and it may be a better start. Simply comment out the "CALL BuildSetupTranslationTransform.cmd xxx" and we have an english installer only.

**Comment by alexhass** [comment]
Please see LibeOffice Documentation how to deploy via AD. You need Orca or InstEd and you can disable all languages except the English 1033.

https://wiki.documentfoundation.org/Deployment_and_Migration

Managed software deployment
This is the recommended method as you can centrally manage which version of LibO is installed on various machines in Active Directory as well as centrally deploy upgrades just by adding a newer package to the GPO.

1. Download the installer for the version of LibO you want to deploy.
2. Open the MSI file in ORCA or InstEd. (See the section on MSI editing for download links.)
3. Go to View ▸ Summary Information... in ORCA or Tables ▸ Summary Info... in InstEd.
4. Remove some of the language codes listed in the Languages field, making sure to keep at least the one(s) you need. http://www.science.co.il/language/locale-codes.asp (This is required to avoid an error when adding the package to the GPO.)
5. Save the MSI to a network location where all machine accounts have Read access.

**Comment by alexhass** [comment]
Has the project the ability to sign the MSI?

**Comment by GitMensch** [comment]
The project has the ability (the worker have `signtool` and the password can be used as secure token (if it isn't set then no signing would happen). But so far no one showed up that sponsored a digital code certificate we can use (issue concerning code-signing: #174).

**Comment by bluikko** [comment]
What would be the cost for the certificate? A quick search shows some vendors ask $300-$500 for a non-EV code signing certificate. Does the project have the ability to accept donations if I can get the company to donate that money? I can't promise anything at this point but I can try at least.

**Comment by GitMensch** [comment]
@stripedpajamas "Does the project have the ability to accept donations [...]?"
I don't know. But something that may work much easier is ask the vendors directly to create the certificate for the project an pay in in behalf of it as a company. https://comodosslstore.com/codesigning.aspx  would be 319 USD for one year and 798 for three years.

There's an option for signing as part of a CI action "for free" at https://about.signpath.io/open-source - @daiyam @stripedpajamas would this be reasonable to use?

**Comment by bluikko** [comment]
https://shop.globalsign.com/en/code-signing is $289 (1 year) / $599 (3 years) - if a paid certificate is acquired it pretty much does not matter where it's from and the cheapest could be OK? I'd expect any CA selling code signing certificates to be pretty universally accepted?

**Comment by GitMensch** [comment]
> I'd expect any CA selling code signing certificates to be pretty universally accepted?

Not completely, it depends where their root certificate is installed (some older/newer Windows versions may not have them installed).

> the cheapest could be OK.

Could be, depends on the need. To not have Windows Smart Screen be shown at all - like the VSCode installer - you'd need an "extended validation" one, but I'd have to check if you can put this into a CI pipeline.

**Comment by daiyam** [comment]
Let's keep it active.

**Comment by bluikko** [comment]
> Can we add an English language last to possibly work around that?

It would be a great improvement IMHO to set the language 1033 (`en-us`? or would it be specified as 1033? I'm sure you know better) as last: it would mean the majority of the users (who likely have the 1033 language) would not need to specifically change something to install the package in cases where package language matters.  
I personally at least often forget to set the "ignore package language" flag when adding a new version and then nobody gets the package until I remove&readd with the flag set - this option is something that was not mentioned in https://github.com/VSCodium/vscodium/issues/830#issuecomment-916726209 and the "disable languages" workaround is needed only in special cases (cases where some data in the MSI was too large IIRC).  

I have also seen MSI packages with "unspecified" as the language - but I do not know what this means in practice or how that's achieved; or indeed if it's relevant here at all when languages are included in the MSI.

**Comment by daiyam** [comment]
@GitMensch I've contacted SignPath. So it seems that we can the windows binaries to be signed by them but it would require to have the build process to AppVeyor because they have an automatic vetting process for the binary artifacts from OSS projects.

**Comment by GitMensch** [comment]
AppVeyor can be _really_ slow; wouldn't integration with GitHub actions be useful for them, too?
If they don't - do you have an idea how long the run may take on Appveyor?

Note: If this helps "a lot" I could try to setup a appveyor configuration for the window builds (would be x86 x64 and arm64 via msys2).

**Comment by daiyam** [comment]
> wouldn't integration with GitHub actions be useful for them, too?

I did ask. 

> If they don't - do you have an idea how long the run may take on Appveyor?

No idea. I don't know AppVeyor. So any help would be welcome if we go that route.

**Comment by GitMensch** [comment]
Will do, so assigned to me until ready, then will assign it to you afterwards.
